filesystem/block.go

```diff
@@ -0,0 +1,340 @@
+	}
+
+	return nil
+}
+
+// grow resizes the data to a larger length.
+// Requirement: to > from.
+// If the file ended in a partial block, we fetch it, grow it,
+// and write it back.
+func grow(e sqlExecutor, inodeID, from, to uint64) error {
+	addRange := newBlockRange(from, to-from)
+	insertFrom := addRange.start
+
+	if addRange.startOffset > 0 {
+		// We need to extend the original 'last block'.
+		// Fetch it, grow it, and update it.
+		// TODO(marc): this would be more efficient if we had RPAD for bytes.
+		data, err := getBlockData(e, inodeID, addRange.start)
+		if err != nil {
+			return err
+		}
+		data = append(data, make([]byte, addRange.startLength, addRange.startLength)...)
+		if err := updateBlockData(e, inodeID, addRange.start, data); err != nil {
+			return err
+		}
+		// We don't need to insert this block.
+		insertFrom++
+	}
+
+	insertTo := addRange.last
+	if insertTo < insertFrom {
+		return nil
+	}
+
+	// Build the sql statement and blocks to insert.
+	// We don't share this functionality with 'write' because we can repeat empty blocks.
+	// This would be shorter if we weren't trying to be efficient.
+	// TODO(marc): this would also be better if we supported sparse files.
+	paramStrings := []string{}
+	params := []interface{}{}
+	count := 1 // placeholder count starts at 1.
+	if insertFrom != insertTo {
+		// We have full blocks. Only send a full block once.
+		params = append(params, make([]byte, BlockSize, BlockSize))
+		count++
+	}
+
+	// Go over all blocks that are certainly full.
+	for i := insertFrom; i < insertTo; i++ {
+		paramStrings = append(paramStrings, fmt.Sprintf("(%d, %d, $1)", inodeID, i))
+	}
+
+	// Check the last block.
+	if addRange.lastLength > 0 {
+		// Not empty, write it. It can't be a full block, because we
+		// would have an empty block right after.
+		params = append(params, make([]byte, addRange.lastLength, addRange.lastLength))
+		paramStrings = append(paramStrings, fmt.Sprintf("(%d, %d, $%d)",
+			inodeID, addRange.last, count))
+		count++
+	}
+
+	if len(paramStrings) == 0 {
+		// We had only one block, and it was empty. Nothing do to.
+		return nil
+	}
+
+	insStmt := fmt.Sprintf(`INSERT INTO fs.block VALUES %s`, strings.Join(paramStrings, ","))
+	if _, err := e.Exec(insStmt, params...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// read returns the data [from, to).
+// Requires: to > from and [to, from) is contained in the file.
+func read(e sqlExecutor, inodeID, from, to uint64) ([]byte, error) {
+	readRange := newBlockRange(from, to-from)
+	end := readRange.last
+	if readRange.lastLength == 0 {
+		end--
+	}
+
+	blockInfos, err := getBlocksBetween(e, inodeID, readRange.start, end)
+	if err != nil {
+		return nil, err
+	}
+	if len(blockInfos) != end-readRange.start+1 {
+		return nil, fmt.Errorf("wrong number of blocks, asked for [%d-%d], got %d back",
+			readRange.start, end, len(blockInfos))
+	}
+
+	if readRange.lastLength != 0 {
+		// We have a last partial block, truncate it.
```

Comment:
the partial block here refers the blocks we want to read, `readRange` is initialized from the read offset and length.
This just means that we want to read X full blocks plus a partial block. This last partial bit may come from a full block (eg: `read(10blockfile, 0, 1block + 1`)` so we need to grab only part of it.